[perfetto] initial port of 43.1

rtzoeller opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

rtzoeller committed:
[perfetto] initial port of 43.1

Co-authored-by: WangWeiLin-MV <156736127+WangWeiLin-MV@users.noreply.github.com>

rtzoeller commented:
@microsoft-github-policy-service agree company="ALIARO"

rtzoeller commented:
I've only directly tested the output on Linux x64, but I have no reason to believe it will not work in other environments.

dg0yt commented:
How much overlap has this port with an official (`gn`) build? 
Isn't it possible to use a shared lib on non-windows?

rtzoeller commented:
> How much overlap has this port with an official (gn) build?

@dg0yt one of the outputs of the official build is the pair of amalgamated source files we're consuming, which are attached to each tagged release (e.g. [for 43.0](https://github.com/google/perfetto/commit/355f5d8074c86b037bca5d0b4588e8e3e2a93cdd)).

This greatly simplifies the build process, and matches the [example built with CMake](https://github.com/google/perfetto/blob/main/examples/sdk/CMakeLists.txt).

>  Isn't it possible to use a shared lib on non-windows?

Maybe, but I've not tested it here and presumed we could enable it later if needed without breaking compatibility.

If this is a blocker for the initial port let me know and I'll happily investigate.

rtzoeller commented:
> Isn't it possible to use a shared lib on non-windows?

Looks like it is. I'll get a fix pushed shortly.

dg0yt commented on ports/perfetto/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+include(CMakePackageConfigHelpers)
+
+include_directories(sdk)
+add_library(perfetto sdk/perfetto.cc)
```
I think we still need
```suggestion
add_library(perfetto sdk/perfetto.cc)
target_include_directories(perfetto PUBLIC $<LINK_INTERFACE:include>)
```
~~~

rtzoeller replied:
Feedback should all be addressed.

dg0yt commented on ports/perfetto/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+
+install(
+    EXPORT unofficial-perfetto-config
+    NAMESPACE unofficial::
```
The suggested pattern is:
```suggestion
    NAMESPACE unofficial::perfetto::
```

dg0yt commented on ports/perfetto/usage:
```diff
@@ -0,0 +1,4 @@
+The package perfetto provides an unofficial CMake target:
+
+    find_package(unofficial-perfetto CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE unofficial::perfetto)
```
I wouldn't install a file if the heuristics output is good enough.
And for consistency, it uses a slightly different wording and formatting:
~~~
perfetto provides CMake targets:

  find_package(unofficial-perfetto CONFIG REQUIRED)
  target_link_libraries(main PRIVATE unofficial::perfetto::perfetto)
~~~

rtzoeller replied:
Ah, at one point it wasn't generating any output automatically for me and so I added the file manually. Looks like that was probably due to some other issue in my configuration at the time.

Thanks!

WangWeiLin-MV commented on ports/perfetto/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+cmake_minimum_required(VERSION 3.8.0)
+project(perfetto)
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+include(CMakePackageConfigHelpers)
+
+add_library(perfetto sdk/perfetto.cc sdk/perfetto.h)
+target_include_directories(perfetto PUBLIC $<INSTALL_INTERFACE:include>)
+
+if (MSVC)
+    target_compile_options(perfetto PRIVATE "/bigobj")
+    target_compile_options(perfetto PRIVATE "/permissive-")
+    target_compile_definitions(perfetto PRIVATE WIN32_LEAN_AND_MEAN NOMINMAX)
+endif (MSVC)
+
+install(
+    TARGETS perfetto
+    EXPORT unofficial-perfetto-config
+    LIBRARY DESTINATION lib
+    ARCHIVE DESTINATION lib
+)
+
+install(FILES sdk/perfetto.h DESTINATION include/)
+
+install(
+    EXPORT unofficial-perfetto-config
+    NAMESPACE unofficial::perfetto::
+    DESTINATION share/unofficial-perfetto
+)
```
```suggestion
# Unofficial perfetto CMakeLists.txt from https://github.com/google/perfetto/blob/v43.1/meson.build
cmake_minimum_required(VERSION 3.23)
project(perfetto LANGUAGES CXX)

add_library(perfetto)
target_compile_features(perfetto PRIVATE cxx_std_17)
target_sources(perfetto
    PRIVATE "sdk/perfetto.cc"
    PUBLIC FILE_SET HEADERS BASE_DIRS "sdk" FILES "sdk/perfetto.h"
)

set(THREADS_PREFER_PTHREAD_FLAG TRUE)
find_package(Threads REQUIRED)
target_link_libraries(perfetto PRIVATE Threads::Threads)

if(ANDROID)
    target_link_libraries(perfetto PRIVATE log)
endif(ANDROID)

if(WIN32)
    target_compile_options(perfetto PRIVATE "/bigobj")
    target_compile_definitions(perfetto PRIVATE WIN32_LEAN_AND_MEAN NOMINMAX)
    target_link_libraries(perfetto PRIVATE ws2_32)
endif(WIN32)

if(MSVC)
    target_compile_options(perfetto PRIVATE "/permissive-")
endif(MSVC)

install(TARGETS perfetto
    EXPORT unofficial-perfetto-config
    FILE_SET HEADERS DESTINATION "include"
)

install(EXPORT unofficial-perfetto-config
    NAMESPACE unofficial::perfetto::
    DESTINATION "share/unofficial-perfetto"
)

install(FILES "protos/perfetto/trace/perfetto_trace.proto" DESTINATION "share/unofficial-perfetto")
```

WangWeiLin-MV commented on ports/perfetto/portfile.cmake:
```diff
@@ -0,0 +1,25 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO google/perfetto
+    REF v43.1
```
```suggestion
    REF "v${VERSION}"
```

WangWeiLin-MV commented on ports/perfetto/portfile.cmake:
```diff
@@ -0,0 +1,25 @@
+
+vcpkg_cmake_install()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+
+vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-${PORT})
```
```suggestion
vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-${PORT})

file(REMOVE_RECURSE
    "${CURRENT_PACKAGES_DIR}/debug/include"
    "${CURRENT_PACKAGES_DIR}/debug/share"
)
```

rtzoeller commented:
@WangWeiLin-MV I've squashed the proposed changes into the original commit and added you as a co-author for it. This should be ready for another review.

WangWeiLin-MV commented on ports/perfetto/vcpkg.json:
```diff
@@ -0,0 +1,18 @@
+  "description": "System profiling, app tracing and trace analysis",
+  "homepage": "https://perfetto.dev",
+  "license": "Apache-2.0",
+  "supports": "!uwp & !x86",
```
Should `x86` be excluded? The local test passed for triplet `x86-windows`.

rtzoeller replied:
> Should `x86` be excluded? The local test passed for triplet `x86-windows`.

Hmm, it had failed on an earlier version of this PR, which is why I disabled it. That may have been an unrelated failure (I don't recall the message).

I'll be available in a few hours to try re-enabling it to see if the automated PR checks pass.

UWP definitely doesn't work.

rtzoeller replied:
Seems to be passing now, so I'll leave it enabled!

WangWeiLin-MV reviewed (dismissed):
Port usage tests pass with following triplets:
* x64-linux
* x64-windows
* x64-windows-static
* x64-windows-static-md
* x86-windows

JavierMatosD commented:
Hi @rtzoeller,

We generally aim to use upstream's build configurations directly and only resort to providing unofficial exports when integrating with upstream would involve extensive patching or customization that doesn't seem justifiable. Could you share your rationale for opting for an unofficial CMake export instead of adapting Perfetto's existing Meson build for vcpkg? I'm not deeply familiar with Meson and its intricacies, so your insights would be very valuable to me and perhaps highlight aspects I might not be aware of.

Thanks for your contribution and looking forward to your thoughts!

rtzoeller commented:
Thanks for the insight @JavierMatosD.

> Could you share your rationale for opting for an unofficial CMake export instead of adapting Perfetto's existing Meson build for vcpkg?

I wasn't previously aware of any existing Meson support in Perfetto; the primary [build instructions](https://perfetto.dev/docs/contributing/build-instructions) only mention GN as a build system, although the repo also provides [an example](https://github.com/google/perfetto/tree/main/examples/sdk) of using CMake to consume the amalgamated source files.  In my usage of Perfetto I've always consumed the amalgamated SDK source files (e.g. [for v43.0](https://github.com/google/perfetto/commit/355f5d8074c86b037bca5d0b4588e8e3e2a93cdd)), and given the example CMake build this was the most approachable option.

I now see that there is a [`meson.build`](https://github.com/google/perfetto/blob/master/meson.build) file, and can take a look at consuming it.

> I'm not deeply familiar with Meson and its intricacies, so your insights would be very valuable to me and perhaps highlight aspects I might not be aware of.

I have zero familiarity with Meson, so please bear with me as I attempt to rework this PR. :smile:

rtzoeller commented:
At first glance the [existing Meson support](https://github.com/google/perfetto/blob/master/meson.build) appears to be rather incomplete.

AFAICT, it only supports building a static library and does not install any headers (but does include the `.proto` file). It's likely possible for us to patch `meson.build` as part of the port, but I'm unsure that's any better than the current CMake situation -- the minimal [commit history for `meson.build`](https://github.com/google/perfetto/commits/master/meson.build) and lack of documentation/examples using Meson admittedly don't inspire a ton of confidence in me.

---

@JavierMatosD can you please advise on which direction is acceptable for moving forward? Thanks!

- Continue maintaining an unofficial CMake build (i.e. merge the existing PR).
- Patch the existing Meson build system to also install `perfetto.h` to the `include` directory.
  - Optionally patch the Meson build system to support building as a dynamic library.
  - Full disclosure: I have effectively no experience with Meson other than reviewing the existing build just now.
- Move away from the amalgamated SDK builds provided by the official Perfetto releases and build the source directly with GN instead.
  - Full disclosure: I have no experience with GN.

JavierMatosD commented:
Big thanks for diving into the integration options! I'm initially leaning towards the unofficial CMake route because it seems super practical for now. But, I want to run it by our team to double-check we’re all on the same page for the long haul.

I’ll circle back with you as soon as we've made a decision. Really appreciate your patience and all the effort you're putting into this!

dg0yt commented:
This port has no dependencies. So providing an unofficial CMake config written by hand shouldn't be more difficult write by hand than a whole vendored CMake build system. If meson doesn't already provided official pkg-config files.

JavierMatosD commented on ports/perfetto/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+    PUBLIC FILE_SET HEADERS BASE_DIRS "sdk" FILES "sdk/perfetto.h"
+)
+
+set(THREADS_PREFER_PTHREAD_FLAG TRUE)
```
Should `PThreads` be a dependency (or feature) in the `vcpkg.json`?

rtzoeller replied:
Seems like it, yes. I'll add it.

JavierMatosD commented on ports/perfetto/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+
+if(WIN32)
+    target_compile_options(perfetto PRIVATE "/bigobj")
+    target_compile_definitions(perfetto PRIVATE WIN32_LEAN_AND_MEAN NOMINMAX)
```
Why are these preprocessor definitions needed? (`WIN32_LEAN_AND_MEAN`, `NOMINMAX`)

rtzoeller replied:
`NOMINMAX` is needed to prevent the definition of macros which conflict with perfetto's use of `std::min` and `std::max`.

`WIN32_LEAN_AND_MEAN` is needed to compile both `winsock.h` and `ws2def.h` successfully.

```
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(103): warning C4005: 'AF_IPX': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(457): note: see previous definition of 'AF_IPX'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(147): warning C4005: 'AF_MAX': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(476): note: see previous definition of 'AF_MAX'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(187): warning C4005: 'SO_DONTLINGER': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(399): note: see previous definition of 'SO_DONTLINGER'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(240): error C2011: 'sockaddr': 'struct' type redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(482): note: see declaration of 'sockaddr'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(442): error C2143: syntax error: missing '}' before 'constant'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(442): error C2059: syntax error: 'constant'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(496): error C2143: syntax error: missing ';' before '}'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(496): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(496): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(577): warning C4005: 'IN_CLASSA': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(284): note: see previous definition of 'IN_CLASSA'

```

JavierMatosD commented on ports/perfetto/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+endif(WIN32)
+
+if(MSVC)
+    target_compile_options(perfetto PRIVATE "/permissive-")
```
Why is this needed?

rtzoeller replied:
To be clear that turns _off_ permissive mode (note the minus), and puts MSVC into standards-compliant mode.

I wasn't able to get it to fail without this flag as part of the automated build, but my recollection from previous usage is that it's needed for older MSVC versions  (2017?). 

It's in the [recommended set of flags](https://perfetto.dev/docs/instrumentation/tracing-sdk) for using the tracing compiler:

```cmake
# Enable standards-compliant mode when using the Visual Studio compiler.
if (MSVC)
  target_compile_options(example PRIVATE "/permissive-")
endif (MSVC)
```

JavierMatosD replied:
If it's not strictly required we should remove this. Thanks for the explanation!

JavierMatosD commented:
Setting the PR to draft. Please mark ready for review when you've responded. That way I will know you've responded

rtzoeller commented on ports/perfetto/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+project(perfetto LANGUAGES CXX)
+
+add_library(perfetto)
+target_compile_features(perfetto PRIVATE cxx_std_17)
```
Yes, see the [v33.1 release notes](https://github.com/google/perfetto/releases/tag/v33.1).

> Switched to a C++17-only project by removing C++11 opt-out. This completes the migration started in v31.0.

rtzoeller commented on ports/perfetto/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+)
+
+set(THREADS_PREFER_PTHREAD_FLAG TRUE)
+find_package(Threads REQUIRED)
```
I'm unsure what you mean here aside from what's asked in the other comment.

> Should PThreads be a dependency (or feature) in the vcpkg.json?

WangWeiLin-MV replied:
The lib `pthread` is a system library, it just requires the `find_package`, and it will only be used on NON-WINDOWS. The checks in upstream https://github.com/google/perfetto/blob/v43.1/include/perfetto/ext/base/thread_checker.h#L22-L24.

JavierMatosD commented on ports/perfetto/vcpkg.json:
```diff
@@ -0,0 +1,19 @@
+  "license": "Apache-2.0",
+  "supports": "!uwp & !x86",
+  "dependencies": [
+    "pthreads",
```
```suggestion
    {
         "name"" "pthreads",
         "platform": "!windows"
    },
```

dg0yt replied:
@JavierMatosD Are you serious about this? Almost no port does this, but almost every non-trivial port uses pthreads.

JavierMatosD replied:
Serious about what exactly? Adding the platform expression or adding it as a dependency?

dg0yt replied:
Do we really request a `pthreads` dependency on platforms where the port is empty and where most ports don't use the dependency and where depending ports don't need to be rebuilt when the windows-only implementation is updated?
